### ISSUE TITLE
Blank black screen when opening EPUB book

### ISSUE THREAD

**Issue by plotn** [koreader/koreader#8583]
* KOReader version: 2021.12.1
* Device: Xiaomi Mi 9 (cepheus) 

Hi! I recently downloaded Rust book repository: https://github.com/ruRust/rust_book_2ed
Then I compiled it via mdbook, then via mdbook-epub (https://github.com/Michael-F-Bryan/mdbook-epub). The file is attached.
So neither KOReader not CoolReader can open the book. Alreader X, Librera - the same. Only MoonReader was able.

</details>
[The Rust Programming Language.zip](https://github.com/koreader/koreader/files/7771352/The.Rust.Programming.Language.zip)

**Comment by plotn** [comment]
[The Rust Programming Language.zip](https://github.com/koreader/koreader/files/7771364/The.Rust.Programming.Language.zip)

**Comment by Frenzie** [comment]
The document is quite severely broken.

[epubcheck.log](https://github.com/koreader/koreader/files/7771400/epubcheck.log)

The most basic surgery on the document (redoing the zipfile without duplicate references) doesn't seem to do a lot of good but the resulting floating point error seems like something we might want to avoid if it interests @poire-z  
[The Rust Programming Language-duplicates-removed.zip](https://github.com/koreader/koreader/files/7771416/The.Rust.Programming.Language-duplicates-removed.zip)

**Comment by poire-z** [comment]
The floating point exception (that I get only when re-opening the document from its empty cache file) can be solved with:
```diff
--- a/crengine/include/lvhashtable.h
+++ b/crengine/include/lvhashtable.h
    LVHashTable( int size )
    {
        if (size < 16 )
            size = 16;
        _table = new pair* [ size ]();
        _size = size;
        _count = 0;
    }
[...]
     int size() { return _size; }
     void resize( int nsize )
     {
+        if (nsize < 16 )
+            nsize = 16;
         pair ** new_table = new pair * [ nsize ]();
                if (_table) {
                        for ( int i=0; i<_size; i++ ) {
```

For the black page = no content, I think it's because the XHTML fragments in the EPUBs are not proper XHTML files:
(pardon my cyrillic :):
```html
<h1>¿¿¿¿¿¿¿¿</h1>
<p>¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿! ¿¿¿ ¿¿¿¿¿ ¿ ¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿¿¿¿¿¿¿ Rust.
Rust - ¿¿¿ ¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿¿¿¿¿¿¿, ¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿ ¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿,
¿¿¿¿¿¿¿ ¿ ¿¿¿¿¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿. ¿¿¿¿¿¿¿¿¿¿¿¿, ¿¿¿¿¿¿¿¿ ¿ ¿¿¿¿¿¿¿¿¿¿¿¿¿¿¿ ¿
[...]
<p>¿¿¿¿ ¿¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿, ¿¿¿¿¿¿¿¿¿¿, ¿¿¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿¿ ¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿¿¿¿
<a href="https://github.com/ruRust/rust_book_2ed">GitHub</a>. <a href="https://github.com/ruRust/rust_book_2ed/blob/ru_version/CONTRIBUTING.md">CONTRIBUTING.md</a> ¿ ¿¿¿ ¿¿¿¿, ¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿
¿¿¿¿¿¿¿¿ ¿¿¿¿¿¿¿¿¿.</p>
```
Add `<html><body>` at the top and `</body></html>` at the bottom of one of these html files, and you get some content displayed.

It's so probably not per EPUB specs to not have proper XHTML fragments - but I dunno if we can/should losen the HTML detection and trust the `media-type="application/xhtml+xml"` in the opf : 
`<item media-type="application/xhtml+xml"  id="ch01-00-introduction_html" href="ch01-00-introduction.html" />
`
I think the code expects to see a HTML node and a BODY node in the HTML fragments to do some substitutions (html > DocFragment) and CSS insertions - so it's probably not an easy thing to solve. Probably not worth complexifying our code trying to be one of the few readers that could open it, when most of the other readers don't :)

**Comment by plotn** [comment]
I agree )))
It is strange that document was generated so broken, but - this story is at the side of those tool (mdbook-epub) ...

**Comment by Frenzie** [comment]
@poire-z 
> I think the code expects to see a HTML node and a BODY node in the HTML fragments to do some substitutions (html > DocFragment) and CSS insertions - so it's probably not an easy thing to solve. Probably not worth complexifying our code trying to be one of the few readers that could open it, when most of the other readers don't :)

They're optional in regular HTML, but required in XHTML, right? So yeah, I doubt we should change that.

@plotn 
> It is strange that document was generated so broken, but - this story is at the side of those tool (mdbook-epub) ...

It generates the EPUB straight from Markdown? In that case it's definitely an issue with the tool. But if there was something in between, you can imagine that a Markdown to HTML converter has two settings. One where it only converts the Markdown to HTML, and one where it converts a Markdown file into a full HTML document.

**Comment by poire-z** [comment]
> They're optional in regular HTML, but required in XHTML, right? So yeah, I doubt we should change that.

I believe so.
Although the latest EPUB specs say something like: we no more specify the html bits, follow the "living html standard" specs: https://github.com/koreader/koreader/issues/7219#issuecomment-770437164
- but that issue says it won't be easy to support unbalanced html properly. Fixing missing html and body would only be a subset of that, and non essential.
So yes, not changing that for now.

**Comment by plotn** [comment]
There were nothing "in between", only "applying the tool" - so it on it's side.

By the way, i found another one reader, that could read it - F3 from Total Commander (i dont know what plugin actually does it).
Calibre also opens it without any problem, and calibre epub editor has beautiful utility - "fix html files" - it added <HTML><BODY> in each file, so it successfully open in "crengine-based" readers.

**Comment by Frenzie** [comment]
I don't think that XML (and as such XHTML5) has implicit elements. The living HTML standard thing would therefore only refer to elements like `<aside>` that are new in HTML5.

https://html.spec.whatwg.org/multipage/xhtml.html#xhtml

**Comment by poire-z** [comment]
When I read around at the time, it felt like the latest EPUB specs were saying: go with HTML5, no longer need for balanced XHTML :)
> Rules for text/html resources are discussed in the section above entitled "The HTML syntax".

Or it depends on the mimetype in the opf, where both text/xhtml and text/html are allowed.
In which case, this book is still wrong as it announces `<item media-type="application/xhtml+xml"`

**Comment by Frenzie** [comment]
> When I read around at the time, it felt like the latest EPUB specs were saying: go with HTML5, no longer need for balanced XHTML :)

Well, it's still technically not valid HTML either. It should start with `<!DOCTYPE html>`. Few parsers will make much of a problem out of that, but you will trigger quirks mode which is not normally what you want.

**Comment by poire-z** [comment]
Closing as "won't fix".
But the floating point crash is fixed with #8700.
